File: context_os_runtime/approval.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path

from .events import read_events


@dataclass(slots=True)
class ActionStatus:
    final_status: str
    executable: bool
    blacklisted: bool


def derive_action_status(log_path: Path, *, session_id: str, action_hash: str) -> ActionStatus:
    now = datetime.now(UTC)
    requested_event: dict[str, object] | None = None
    approved = False
    final_status = "NOT_ACTIONABLE"
    for event in read_events(log_path):
        if event.get("session_id") != session_id or event.get("action_hash") != action_hash:
            continue
        if event["event_type"] == "ACTION_REQUESTED":
            requested_event = event
            final_status = "PENDING"
        elif event["event_type"] == "HUMAN_APPROVAL_DENIED":
            final_status = "DENIED"
        elif event["event_type"] == "SYSTEM_AUTO_REJECTED":
            final_status = "EXPIRED"
        elif event["event_type"] == "HUMAN_APPROVAL_RECEIVED" and final_status == "PENDING":
            approved = True
            final_status = "APPROVED"

    if requested_event is not None and final_status == "PENDING":
        expires_at = datetime.fromisoformat(str(requested_event["expires_at"]))
        if now > expires_at:
            final_status = "EXPIRED"

    blacklisted = final_status in {"DENIED", "EXPIRED"}
    executable = approved and not blacklisted
    return ActionStatus(final_status=final_status, executable=executable, blacklisted=blacklisted)

File: src/core/approval.ts
import { readEvents } from './event-log';
import type { Event } from './events';

export interface ActionStatus {
  final_status: 'NOT_ACTIONABLE' | 'PENDING' | 'APPROVED' | 'DENIED' | 'EXPIRED';
  executable: boolean;
  blacklisted: boolean;
}

function eventValue(event: Event, key: string): unknown {
  if (event.payload && key in event.payload) return event.payload[key];
  return (event as unknown as Record<string, unknown>)[key];
}

export function deriveActionStatus(
  logPath: string,
  args: { sessionId: string; actionHash: string },
): ActionStatus {
  const now = new Date();
  let requestedEvent: Event | null = null;
  let approved = false;
  let finalStatus: ActionStatus['final_status'] = 'NOT_ACTIONABLE';

  for (const event of readEvents(logPath)) {
    if (event.session_id !== args.sessionId) continue;
    if (eventValue(event, 'action_hash') !== args.actionHash) continue;

    switch (event.event_type) {
      case 'TOOL_REQUESTED':
        requestedEvent = event;
        finalStatus = 'PENDING';
        break;
      case 'TOOL_DENIED': {
        const reason = String(eventValue(event, 'reason') ?? '');
        finalStatus = reason === 'auto_rejected_ttl_expired' ? 'EXPIRED' : 'DENIED';
        break;
      }
      case 'TOOL_APPROVED':
        if (finalStatus === 'PENDING') {
          approved = true;
          finalStatus = 'APPROVED';
        }
        break;
    }
  }

  if (requestedEvent && finalStatus === 'PENDING') {
    const expiresAt = new Date(String(eventValue(requestedEvent, 'expires_at')));
    if (now > expiresAt) finalStatus = 'EXPIRED';
  }

  const blacklisted = finalStatus === 'DENIED' || finalStatus === 'EXPIRED';
  const executable = approved && !blacklisted;
  return { final_status: finalStatus, executable, blacklisted };
}
